Articles/Blog/TeamExperiencesImplementingCI.md

```diff
@@ -0,0 +1,82 @@
+
+To understand what has worked best, this article interviews three development teams across the DOE Exascale Computing Project that have adopted CI: ExaSMR, Visit, and ZFP.  The ExaSMR project develops simulations of small modular reactors that provide the engineering community with highly detailed, benchmark datasets.  The Visit project is a highly scalable, turnkey application for data analysis, visualization, and movie making.  ZFP is a compact library for lossy compression of multi-dimensional arrays of integer and floating-point data with controllable rate, precision, and accuracy.
+
+Each of these teams has a unique story about how CI fits into their overall project goals, the evolution of their pipelines, and of the challenges they faced along the way.  The adopters of CI have all described its value to be “obvious” in retrospect because of the numerous occasions where either testing or pull-request reviews have caught errors that would have gone undetected otherwise.  ZFP put significant effort into developing tests, resulting in over 5,000 unit tests with nearly full code coverage.  VisIt’s extensive nightly regression testing combined with a permissive policy regarding breaking the mainline of development meant that the project’s initial CI testing could be rather modest and fit within the time limits of freely available services. The project has seen a greater impact from formal reviews of pull requests.  ExaSMR noted that having CI in place sets the expectation that new contributions come with corresponding tests.  This approach results in better quality code contributions from the community that are less stressful for the maintainers to review.
+
+### Case 1: ExaSMR
+
+ExaSMR describes their CI experience with the [OpenMC package](https://github.com/openmc-dev/openmc).  This package simulates Monte Carlo neutron and photon transport for use within multiphysics calculations coupled to computational fluid dynamics.  OpenMC is a C++ library built on top of 3 geometry and math libraries (DAGMC, GSL, xtensor), and 3 I/O libraries (HDF5, pugixml, {fmt} a.k.a. fmtlib), and building both an executable and a modular Python ctypes interface.  The project was first hosted on GitHub in 2012, and Travis CI was first added in February of 2015.  The team used CI to install system packages and run its Python test harness.  In December 2020, the project switched to GitHub actions, which supported the automation features the project had adopted by that time: a matrix of compilation and test options (turning on/off MPI/OpenMP/compiler optimizations), coveralls.io code-coverage integration, and automated creation of docker images for releases.
+
+[Paul Romano](https://github.com/paulromano) noted it was surprisingly easy for the team to get a first useful CI pipeline working with Travis CI. Switching  to GitHub Actions was also relatively straightforward.
```

Comment:
@frobnitzem, here and in two other cases below, you are hyperlinking a person's name.  But these links have nothing to do with the content of the article.  I'd like to remove the links.  Is there are reason we really need them?

If you want the link as a way of giving credit, I think it might be more appropriate to name them in the acknowledgment section and put the links there.

Reply:
I have moved these links to the acknowledgment section.  We should check with the authors again that they are OK with this.